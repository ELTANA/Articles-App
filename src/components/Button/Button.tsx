import { FC, forwardRef } from 'react';
import type { IButtonProps } from './Button.types';
import Spinner from '$svgs/spinner';

const Button: FC<IButtonProps> = forwardRef<HTMLButtonElement, IButtonProps>(
  ({ loading = false, text, className, ...rest }, ref) => {
    return (
      <button
        ref={ref}
        {...rest}
        className={`text-white bg-blue-500 border-0 py-3 px-6 focus:outline-none hover:bg-blue-600 rounded text-base md:text-lg inline-flex gap-3 items-center font-medium disabled:bg-opacity-40 ${className}`}
      >
        {loading ? <Spinner className="animate-spin h-5 w-5 text-white" /> : null}
        {text}
      </button>
    );
  }
);

Button.displayName = 'Button';

export default Button;
