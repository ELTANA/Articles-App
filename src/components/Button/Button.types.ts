import type { HTMLAttributes } from 'react';

export interface IButtonProps extends HTMLAttributes<HTMLButtonElement> {
  loading?: boolean;
  text: string;
}
