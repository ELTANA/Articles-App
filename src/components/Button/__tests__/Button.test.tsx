import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import Button from '../Button';

describe('Button component', () => {
  test('renders button with text', () => {
    render(<Button text="Click me" />);
    const buttonElement = screen.getByText('Click me');
    expect(buttonElement).toBeInTheDocument();
  });

  test('handles click event', () => {
    const handleClick = jest.fn();
    render(<Button text="Click me" onClick={handleClick} />);
    const buttonElement = screen.getByText('Click me');
    fireEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('renders spinner when loading is true', () => {
    render(<Button text="Click me" loading />);
    const spinnerElement = screen.getByTestId('spinner');
    expect(spinnerElement).toBeInTheDocument();
  });

  test('applies custom class name', () => {
    render(<Button text="Click me" className="custom-class" />);
    const buttonElement = screen.getByText('Click me');
    expect(buttonElement).toHaveClass('custom-class');
  });

  test('applies additional props', () => {
    render(<Button text="Click me" data-testid="custom-button" />);
    const buttonElement = screen.getByTestId('custom-button');
    expect(buttonElement).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const { asFragment } = render(<Button text="Click me" />);
    expect(asFragment()).toMatchSnapshot();
  });
});
