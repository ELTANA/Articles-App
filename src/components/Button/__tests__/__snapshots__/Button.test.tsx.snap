// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="text-white shadow-lg bg-blue-500 border-0 py-3 px-6 focus:outline-none hover:bg-blue-600 rounded text-base md:text-lg inline-flex gap-3 items-center font-medium disabled:bg-opacity-40 undefined"
  >
    Click me
  </button>
</DocumentFragment>
`;
